content/4_Gold/BFS.mdx

```diff
@@ -576,6 +576,134 @@ int main() {
 
 </CPPSection>
 
+<JavaSection>
+
+```java
+import java.util.*;
+import java.io.*;
+
+public class tracks {
+    static final int[] dx = {0, 0, -1, 1};
+    static final int[] dy = {-1, 1, 0, 0};
+    static int N = 1, H, W;
+    static int[][] grid;
+    static int[][] count;
+    public static void main(String[] args) throws IOException {
+        FastIO io = new FastIO();
+
+        H = io.nextInt();
+        W = io.nextInt();
+        grid = new int[H][W];
+
+        for (int i = 0; i < H; i++) {
+            String line = io.next();
+            for (int j = 0; j < W; j++) {
+                grid[i][j] = (line.charAt(j) == 'F')?1:(line.charAt(j) == 'R')?2:-1;
+            }
+        }
+
+        io.println(bfs());
+        io.close();
+    }
+
+    private static int bfs() {
+        count = new int[H][W];
+
+        LinkedList<int[]> q = new LinkedList<>();
+        q.add(new int[]{0,0});
+        count[0][0] = 1;
+
+        while (!q.isEmpty()) {
+            int[] curr = q.poll();
+
+            N = Math.max(N, count[curr[0]][curr[1]]);
+
+            for (int i = 0; i < 4; i++) {
+                int nx = curr[0] + dx[i];
+                int ny = curr[1] + dy[i];
+
+                if(nx < 0 || ny < 0 || nx >= H || ny >= W) continue;
+                if(count[nx][ny] > 0) continue;
+                if(grid[nx][ny] == -1) continue;
+                if(grid[nx][ny] != grid[curr[0]][curr[1]]) {
+                    count[nx][ny] = count[curr[0]][curr[1]] + 1;
+                    q.addLast(new int[]{nx, ny});
+                }
+                else {
+                    count[nx][ny] = count[curr[0]][curr[1]];
+                    q.addFirst(new int[]{nx, ny});
+                }
+            }
+        }
+
+        return N;
+    }
+
+    private static class FastIO extends PrintWriter {
```

Comment:
just asking, is fastio 100% needed for this? if not, then it's probably best not to add it (or at least wrap it in a snippet)?

Reply:
It will definitely TLE without it bc it TLEs with it. Ig I'll wrap it in a snippet.

Reply:
oh yeah sorry i didn't see the first comment